tensorflow/core/kernels/data/sampling_dataset_op.cc

```diff
@@ -0,0 +1,202 @@
+                   DatasetBase** output) override {
+    float rate;
+    OP_REQUIRES_OK(ctx, ParseScalarArgument<float>(ctx, "rate", &rate));
+    *output = new Dataset(ctx, rate, input);
+  }
+
+ private:
+  class Dataset : public DatasetBase {
+   public:
+    Dataset(OpKernelContext* ctx, float rate, const DatasetBase* input)
+        : DatasetBase(DatasetContext(ctx)), rate_(rate), input_(input) {
+      input_->Ref();
+    }
+
+    ~Dataset() override { input_->Unref(); }
+
+    std::unique_ptr<IteratorBase> MakeIteratorInternal(
+        const string& prefix) const override {
+      if (rate_ == 0) {
+        return std::unique_ptr<IteratorBase>(new EmptyIterator(
+            {this, strings::StrCat(prefix, "::EmptySample")}));
+      } else {
+        return std::unique_ptr<IteratorBase>(
+            new Iterator({this, strings::StrCat(prefix, "::Sampling")}));
+      }
+    }
+
+    const DataTypeVector& output_dtypes() const override {
+      return input_->output_dtypes();
+    }
+
+    const std::vector<PartialTensorShape>& output_shapes() const override {
+      return input_->output_shapes();
+    }
+
+    string DebugString() const override { return "SamplingDatasetOp::Dataset"; }
+
+   protected:
+    Status AsGraphDefInternal(SerializationContext* ctx,
+                              DatasetGraphDefBuilder* b,
+                              Node** output) const override {
+      Node* input_graph_node = nullptr;
+      TF_RETURN_IF_ERROR(b->AddInputDataset(ctx, input_, &input_graph_node));
+      Node* rate = nullptr;
+      TF_RETURN_IF_ERROR(b->AddScalar(rate_, &rate));
+      TF_RETURN_IF_ERROR(b->AddDataset(this, {input_graph_node, rate}, output));
+      return Status::OK();
+    }
+
+   private:
+    class EmptyIterator : public DatasetIterator<Dataset> {
+     public:
+      explicit EmptyIterator(const Params& params)
+          : DatasetIterator<Dataset>(params) {}
+      Status GetNextInternal(IteratorContext* ctx,
+                             std::vector<Tensor>* out_tensors,
+                             bool* end_of_sequence) override {
+        *end_of_sequence = true;
+        return Status::OK();
+      }
+
+     protected:
+      std::shared_ptr<model::Node> CreateNode(
+          IteratorContext* ctx, model::Node::Args args) const override {
+        return model::MakeKnownRatioNode(std::move(args),
+                                         /*ratio=*/1);
+      }
+
+      Status SaveInternal(IteratorStateWriter* writer) override {
+        return Status::OK();
+      }
+
+      Status RestoreInternal(IteratorContext* ctx,
+                             IteratorStateReader* reader) override {
+        return Status::OK();
+      }
+    };
+
+    class Iterator : public DatasetIterator<Dataset> {
+     public:
+      explicit Iterator(const Params& params)
+          : DatasetIterator<Dataset>(params) {
+        std::random_device device("/dev/urandom");
```

Comment:
This is does not make it possible to control the RNG in the kernel through a seed.

Instead, the op should, similar to `RandomDatasetOp`, take `seed` and `seed2` arguments and use them to create `random::PhiloxRandom`. The values for the seed arguments can be copied from the `random_uniform_op` similar to how it is done in https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/grappler/optimizers/data/hoist_random_uniform.cc.

Reply:
seed and seed2 parameter from random_uniform now will be passed to sampling_dataset_op,  and now random generator in sampling_dataset will use PhiloxRandom
https://github.com/tensorflow/tensorflow/pull/25384/commits/fa11069c3e85ffbcae19994bdba09c30eee0a6de#diff-a87bcfeaacbb9697bc5998804322ce6cR148